modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotRestoreProcess.java

```diff
@@ -1227,6 +1273,379 @@ private void finishCacheStart(UUID reqId, Map<UUID, Boolean> res, Map<UUID, Exce
+     * @param snpPath Base snapshot path.
+     * @param cacheIds Restoring cache IDs.
+     * @param incIdx Index of incremental snapshot.
+     */
+    private void restoreIncrementalSnapshot(
+        String snpName,
+        String snpPath,
+        Set<Integer> cacheIds,
+        int incIdx
+    ) throws IgniteCheckedException, IOException {
+        File[] segments = walSegments(snpName, snpPath, incIdx);
+
+        UUID incSnpId = ctx.cache().context().snapshotMgr()
+            .readIncrementalSnapshotMetadata(snpName, snpPath, incIdx)
+            .requestId();
+
+        IncrementalSnapshotFinishRecord incSnpFinRec = readFinishRecord(segments[segments.length - 1], incSnpId);
+
+        if (incSnpFinRec == null)
+            throw new IgniteCheckedException("System WAL records for incremental snapshot wasn't found [id=" + incSnpId + ']');
+
+        CacheStripedExecutor exec = new CacheStripedExecutor(ctx.pools().getStripedExecutorService());
+
+        UUID prevIncSnpId = incIdx > 1
+            ? ctx.cache().context().snapshotMgr().readIncrementalSnapshotMetadata(snpName, snpPath, incIdx - 1).requestId()
+            : null;
+
+        long start = U.currentTimeMillis();
+
+        LongAdder applied = new LongAdder();
+
+        try (WALIterator it = walIter(log, segments)) {
+            IgnitePredicate<GridCacheVersion> txVerFilter = null;
+
+            // Skips applying WAL until base snapshot record.
+            while (it.hasNext()) {
+                WALRecord rec = it.next().getValue();
+
+                if (rec.type() == CLUSTER_SNAPSHOT) {
+                    ClusterSnapshotRecord snpRec = (ClusterSnapshotRecord)rec;
+
+                    if (snpRec.clusterSnapshotName().equals(snpName)) {
+                        if (prevIncSnpId == null)
+                            txVerFilter = txVer -> !incSnpFinRec.excluded().contains(txVer);
+
+                        break;
+                    }
+                }
+            }
+
+            // Apply incremental snapshots.
+            while (it.hasNext()) {
+                WALRecord rec = it.next().getValue();
+
+                if (rec.type() == INCREMENTAL_SNAPSHOT_START_RECORD) {
+                    IncrementalSnapshotStartRecord startRec = (IncrementalSnapshotStartRecord)rec;
+
+                    if (startRec.id().equals(incSnpFinRec.id()))
+                        txVerFilter = v -> incSnpFinRec.included().contains(v);
+                }
+                else if (rec.type() == INCREMENTAL_SNAPSHOT_FINISH_RECORD) {
+                    IncrementalSnapshotFinishRecord finRec = (IncrementalSnapshotFinishRecord)rec;
+
+                    if (finRec.id().equals(prevIncSnpId))
+                        txVerFilter = txVer -> !incSnpFinRec.excluded().contains(txVer);
+                }
+                else if (rec.type() == DATA_RECORD_V2) {
+                    DataRecord data = (DataRecord)rec;
+
+                    DataEntry entry = data.writeEntries().get(0);
```

Comment:
Do we really can filter ALL `DataEntry` based on first item?
Even if this works for now, let's filter `DataEntry` one by one.
It will no hurt performance (AFAIK multi `DataEntry` records exists only on backup nodes when transaction committed), because, `DataRecord` in memory, already.

But it will keep logic correct even after future changes